subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1278 @@
+
+	LOG_DBG("");
+	desc->if0_ncm.iMACAddress = 0;
+	sys_dlist_remove(&data->mac_desc_data->node);
+}   /* usbd_cdc_ncm_shutdown */
+
+
+
+static void *usbd_cdc_ncm_get_desc(struct usbd_class_data *const c_data,
+		const enum usbd_speed speed)
+/**
+ * Get function descriptor based on speed parameter
+ */
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *const data = dev->data;
+
+	if (speed == USBD_SPEED_HS) {
+		return data->hs_desc;
+	}
+
+	return data->fs_desc;
+}   /* usbd_cdc_ncm_get_desc */
+
+
+
+static int cdc_ncm_send(const struct device *dev, struct net_pkt *const pkt)
+/**
+ * Send a network packet to the host
+ */
+{
+	struct cdc_ncm_eth_data *const data = dev->data;
+	struct usbd_class_data *c_data = data->c_data;
+	size_t pkt_len = net_pkt_get_len(pkt);
+	struct net_buf *buf;
+	union xmit_ntb_t *ntb;
+
+	LOG_DBG("len: %d", pkt_len);
+
+	if (pkt_len > sizeof(ntb->data)) {
+		LOG_WRN("Trying to send too large packet, drop");
+		return -ENOMEM;
+	}
+
+	if (!atomic_test_bit(&data->state, CDC_NCM_CLASS_ENABLED)
+	    ||  !atomic_test_bit(&data->state, CDC_NCM_IFACE_UP)) {
+		LOG_INF("Configuration is not enabled or interface not ready %ld", data->state);
+		return -EACCES;
+	}
+
+	buf = cdc_ncm_buf_alloc(cdc_ncm_get_bulk_in(c_data));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate buffer");
+		return -ENOMEM;
+	}
+
+	/*
+	 * create (simple) NTB
+	 */
+	ntb = (union xmit_ntb_t *)buf->data;
+
+	ntb->nth.dwSignature   = sys_cpu_to_le32(NTH16_SIGNATURE);
+	ntb->nth.wHeaderLength = sys_cpu_to_le16(sizeof(struct nth16_t));
+	ntb->nth.wSequence     = sys_cpu_to_le16(++data->tx_sequence);
+	ntb->nth.wNdpIndex     = sys_cpu_to_le16(sizeof(struct nth16_t));
+
+	ntb->ndp.dwSignature   = sys_cpu_to_le32(NDP16_SIGNATURE_NCM0);
+	ntb->ndp.wLength       = sys_cpu_to_le16(sizeof(struct ndp16_t)
+		+ (CFG_CDC_NCM_XMT_MAX_DATAGRAMS_PER_NTB + 1)*sizeof(struct ndp16_datagram_t));
+	ntb->ndp.wNextNdpIndex = 0;
+
+	ntb->ndp_datagram[0].wDatagramIndex  =
+			sys_cpu_to_le16(sys_le16_to_cpu(ntb->nth.wHeaderLength)
+					+ sys_le16_to_cpu(ntb->ndp.wLength));
+	ntb->ndp_datagram[0].wDatagramLength = sys_cpu_to_le16(pkt_len);
+	ntb->ndp_datagram[1].wDatagramIndex  = 0;
+	ntb->ndp_datagram[1].wDatagramLength = 0;
+
+	ntb->nth.wBlockLength =
+			sys_cpu_to_le16(sys_le16_to_cpu(ntb->ndp_datagram[0].wDatagramIndex)
+					+ pkt_len);
+
+	if (net_pkt_read(pkt,
+		ntb->data + sys_le16_to_cpu(ntb->ndp_datagram[0].wDatagramIndex), pkt_len)) {
+		LOG_ERR("Failed copy net_pkt");
+		net_buf_unref(buf);
+		return -ENOBUFS;
+	}
+	/* post: now the complete NTB is written. */
+
+	/* adjust length of net_buf accordingly */
+	net_buf_add(buf, sys_le16_to_cpu(ntb->nth.wBlockLength));
+
+	if (sys_le16_to_cpu(ntb->nth.wBlockLength) % cdc_ncm_get_bulk_in_mps(c_data) == 0) {
+		udc_ep_buf_set_zlp(buf);
+	}
+
+	usbd_ep_enqueue(c_data, buf);
+	k_sem_take(&data->sync_sem, K_FOREVER);
+	net_buf_unref(buf);
```

Comment:
If I understand correctly, this is *unref* and not *free*, so the `usbd_ep_enqueue()` still holds a reference to this buffer, and is likely not freed at this moment.  The other `unref()` are after error management.

The Ethernet API takes care of free-ing (unref-ing) the buffer itself when done.

Does it match how you sees it too?

Reply:
Again taken from ECM ;-)  And it is working too.  Would be very fast become obvious if there is a buffer leak.

Reply:
No doubt! And it also makes sense on the source: the imported part looks all great.

Reply:
yes, it does... almost just inserted on TX side framing and on RX side verification... plus a lot of renaming... plus the extra logic in the update handler... plus the request handler... ;-)